File: sort_colors.py
class Solution:
    # @param A a list of integers
    # @return nothing, sort in place
    def sortColors(self, A):
        length = len(A)
        m = 0
        n = length - 1
        idx = 0
        while idx < n + 1:
            if A[idx] == 0:
                A[m], A[idx] = A[idx], A[m]
                m += 1
                idx += 1
                continue
            if A[idx] == 2:
                A[n], A[idx] = A[idx], A[n]
                n -= 1
                continue
            idx += 1

File: various/ruby/intermediate/sort_colors.rb
# https://leetcode.com/problems/sort-colors/description/
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def sort_colors(nums)
  # r -> w -> b
  r = w = 0
  b = nums.size - 1

  while w <= b
    if nums[w] == 2
      nums[w], nums[b] = nums[b], nums[w]
      b -= 1
    elsif nums[w] == 0
      nums[w], nums[r] = nums[r], nums[w]
      r += 1
      w += 1
    else
      w += 1
    end
  end

  nums
end

p sort_colors([2,1,2,0,1,2,0,1,0])
